src/jailer/src/env.rs

```diff
@@ -546,4 +558,72 @@ mod tests {
         // And changing the umask in the Python integration tests is unsafe because of pytest's
         // process management; it can't be isolated from side effects.
     }
+
+    fn skip_not_found(err: std::io::Error) -> std::io::Result<()> {
+        match err.kind() {
+            std::io::ErrorKind::NotFound => Ok(()),
+            _ => Err(err),
+        }
+    }
+
+    fn get_major(dev: u64) -> u32 {
+        unsafe { libc::major(dev) }
+    }
+
+    fn get_minor(dev: u64) -> u32 {
+        unsafe { libc::minor(dev) }
+    }
+
+    #[test]
+    fn test_mknod_and_own_dev() {
+        use std::os::unix::fs::{FileTypeExt, MetadataExt, PermissionsExt};
+
+        // Create environment with good arguments.
+        let good_arg_vals = ArgVals {
+            node: "1",
+            id: "bd65600d-8669-4903-8a14-af88203add38",
+            exec_file: "/proc/cpuinfo",
+            uid: "1001",
+            gid: "1002",
+            chroot_base: "/",
+            netns: Some("zzzns"),
+            daemonize: true,
+        };
+
+        let arg_parser = build_arg_parser();
+        let mut args = arg_parser.arguments().clone();
+        args.parse(&make_args(&good_arg_vals)).unwrap();
+        let env = Env::new(&args, 0, 0).unwrap();
+
+        // Ensure that path buffers without NULL-termination are handled well.
+        assert!(env.mknod_and_own_dev(b"/some/path", 0, 0).is_err());
+
+        let expected_device_permissions = libc::S_IFCHR | libc::S_IRUSR | libc::S_IWUSR;
+
+        // Ensure TUN device node is created with correct major/minor numbers.
+        let tun = b"/dev/net/tun-test\0";
+        let tun_str = CStr::from_bytes_with_nul(tun).unwrap().to_str().unwrap();
+        fs::remove_file(tun_str).or_else(skip_not_found).unwrap();
+        env.mknod_and_own_dev(tun, DEV_NET_TUN_MAJOR, DEV_NET_TUN_MINOR)
+            .unwrap();
+
+        let metadata = File::open(tun_str).unwrap().metadata().unwrap();
```

Comment:
```suggestion
        let metadata = fs::metadata(tun_str).unwrap();
```
This way you avoid opening the file.